### PR TITLE
refactor package to remove sharp edges for schema authors/users (particularly `@row` / `Row`) and improve API

### LINKED ISSUE #49: `Legolas.Row` type aliases impede ergonomic deprecation cycles
I've had this on the backburner to address for a long time, but am only actually confronting it now because there's an [impending downstream breaking schema change I'd like to land soonish](https://github.com/beacon-biosignals/Onda.jl/issues/119).

Part of the benefit of Legolas.jl is supposed to be that a single package version can actually contain/support definitions for multiple versions of the same schema simultaneously. However, the package also currently encourages users to define type aliases for `Legolas.Row` in a manner that I think ends up at odds with the aforementioned benefit.

Let's say I'm the author of a schema `foo@1`:

```jl
const Foo = @row("foo@1", x::Int)
```

Now let's say I want to introduce a breaking schema change, like forcing `x > 0`. One way to do this is:

```jl
const Foo = @row("foo@1", x::Int)

const Foo2 = @row("foo@2", x::Int = x > 0 ? x : error("bad"))
```

This is annoying as a package maintainer, because you are forced to eventually do one of the following as your package evolves:

1. Don't ever deprecate `Foo` as the alias for `foo@1` row construction. Introduce the alias `Foo1` for `Foo`, and then always have `Foo2`, `Foo3`, etc. for each new schema version. Try to teach your users to always use an integer-appended alias at their callsites. In this case, you end up wishing you would've just started with `Foo1` in the first place.

2. Develop some deprecation strategy by which `Foo` always "eventually" means `foo@<latest>` (I use the term "eventually" because it might generally take multiple package version bumps to achieve a single schema version bump, depending on your approach). The simplest approach here is to just tie breaking schema changes to breaking package changes, but that feels like it defeats Legolas' purported ability to enable packages to seamlessly support multiple schema version simultaneously - an ability which is supposed to *ease* deprecation cycles, not complicate them. Maybe for some special classes of breaking changes it's possible to deprecate things cleanly/automatically, esp. if you're willing to overload `Legolas.Row` construction yourself (which probably should be generally discouraged. In general, though, this option seems even more annoying than option 1 to me.

I'd like a better option here, but I think we might have to change Legolas itself to achieve it.

One proposal: What if we took option 1's approach and baked it into Legolas from the get-go as the "supported best practice"? For example, we could drop `@row(schema_name, fields...)` in favor of `@constructor(schema_name, type_name)`:

```jl
# current way
const MyRow = @row("my-schema@1", x::Int)

# proposed way
@schema("my-schema@1", x::Int)
@constructor("my-schema", MyRow)
```

In this example, `@constructor` would generate a definition like:

```jl
const MyRow{n} = Legolas.Row{Legolas.Schema{Symbol("my-schema"),n}}
```

...and we'd define stuff as necessary to ensure that:

- `MyRow(...)` throws a descriptive error informing the caller that they MUST pass in a version integer
- `MyRow{n}(...)` throws a descriptive error if `n` doesn't correspond to a known schema version

IMO this would result in a world that is similar to Option 1 but more ergonomic/structured for everybody involved.

### PR DESCRIPTION
**Pull request by jrevels**
I pulled on the thread of `Legolas.@row` to resolve #49, and accidentally unravelled the knitted sweater that was Legolas' API. 

To elaborate: the overwhelming number of changes included in this single PR result from a development process that kind of resembles "`x` should be added/removed; hmm, adding/removing `x` immediately renders it obvious that `y` needs to be added/removed, or else the API will be inconsistent or have an obvious hole". Now that I know what the final puzzle looks like all put together, it'd be technically possible to break this up into more incremental PRs, but many of them would still each be breaking and I think it'd actually worsen the churn...but even if it didn't, we don't really have the bandwidth to draw out landing these changes 😅

A lot of this PR's changes stem from the design mini-investigation undertaken [here](https://github.com/beacon-biosignals/Legolas.jl/pull/54#issuecomment-1280080189).

Note that this PR supersedes #52, but is based atop @OTDE's commits from that branch to ensure they also receive contribution assignment for this PR once it lands on `main`.

## changelog

I think the best way to onboard into this PR is probably to go through its new tour.jl, but here are a summary of the changes:

- rm'd `Legolas.Row` in favor of the record types described [in this document](https://gist.github.com/jrevels/fdfe939109bee23566d425440b7c759e)

- `Legolas.Schema` -> `Legolas.SchemaVersion`, and associated name changes to more properly refer to "schema versions" where such changes are warranted. Also, the `schema_` prefix has been dropped from many of the methods that were defined against this type.

- rm'd methods against `::Type{<:Schema}` in favor of methods against `::SchemaVersion` (for similar reasons to why Base methods generally prefer `::Val{n}` over `::Type{Val{n}}`).

- More disciplined/uniform API for handling schema identifiers, including replacing `Legolas.Schema(id::AbstractString)` with `first(parse_identifier(id))`

- `Legolas.@row` has been replaced with `Legolas.@schema` (which simply declares a schema) and `Legolas.@version` (which declares a new version of a schema, and generates all associated record type definitions)

- General overhaul of `tour.jl` and docs to both reflect the new changes and to hopefully provide a smoother / more thorough on-ramp for new users.

- More comprehensive introspection/validation utilities:
    - `required_fields`
    - `declaration`
    - `declared`
    - `find_violation`
    - `complies_with`
    - `accepted_field_type`

- `validate(table, schema)` --> `validate(Tables.schema(table), schema)`. This works better/more generically with `find_violation`/`complies_with`, and just yields more control to the caller in general IMO. Furthermore, this makes it harder for callers to accidentally incorrectly assume that `Legolas.validate` validates *row values*.

- Better-tailored error messages for `Legolas.read`/`Legolas.write` now that those methods don't need to punt to the removed `validate` method

- The new Legolas-generated record types are a bit "dumber" than the former `Legolas.Row`, and its more likely that row tables constructed via these record types will have directly inferrable `Tables.Schema`s. Thus, we've removed a small bit of Legolas' schema-guessing code in a manner that removes a bit of responsibility from `Legolas.read`/`Legolas.write` and makes them more efficient. The tradeoff is that these functions have a higher expectation that callers will provide them with tables whose `Tables.Schema`s are inferrable via `Tables.schema`.

- expected properties of schema version declarations are now better validated/enforced:
    - schema version declarations may only contain one statement per required field
    - schema version declarations must include at least one required field
    - resolves https://github.com/beacon-biosignals/Legolas.jl/issues/51
    - resolves https://github.com/beacon-biosignals/Legolas.jl/issues/53
    - Previously, it was technically possible to alter a preexisting schema version declaration, and for such changes to be automatically reflected in the behavior of that schema version's children within the same Julia session. For example, you could redeclare a preexisting schema version `foo@1` with an additional required field compared to its original declaration, which would de facto cause `Legolas.validate` to require the field for `foo@1`'s child schemas. Since there was little enforcement of actual parent/child field compatibility in the first place, it was possible to redefine schema versions in such a manner that rendered them incompatible with their preexisting children. Now that we seek to better enforce parent/child schema version compatibility, we should strive to disallow this kind of accidental invalidation. We achieve this by bluntly disallowing the alteration of preexisting schema version declarations via redeclaration entirely, because a) there's not currently a mechanism for validating parent declarations against preexisting child declarations and b) the ability to alter preexisting schema version declarations is of questionable utility in the first place.
    - Enforcing the previous points also de facto enforces that a child can't accidentally be declared as its own parent

- resolves https://github.com/beacon-biosignals/Legolas.jl/issues/49

- obviates https://github.com/beacon-biosignals/Legolas.jl/issues/50

- resolves https://github.com/beacon-biosignals/Legolas.jl/issues/41

- resolve https://github.com/beacon-biosignals/Legolas.jl/issues/12

- more comprehensive unit tests (though I'm sure there are more test cases we should add more over time)

- bumps minimum supported Julia version to 1.6, which is the latest LTS minor version, just to avoid maintenance overhead  of dealing with Julia 1.3 compatibility. If any user actually requires Julia 1.3 support, please speak up and I'll reconsider.

## Breaking Change Summary

- The removal/replacement of `Legolas.Row` in general. Note that, unless we add a deprecation path, the new Legolas version will not be able to automatically deserialize values that were previously serialized via the former `Legolas.Row` ArrowTypes overloads. Note though that this probably doesn't affect most Legolas-serialized data, unless e.g. you had a nested `Legolas.Row` value inside of a column of a table.
- `Legolas.@row` -> `Legolas.@schema` + `Legolas.@version`
- `Legolas.Schema` -> `Legolas.SchemaVersion`, and associated method name changes
- methods against `Type{<:Schema}` -> methods against `::SchemaVersion`
- `Schema(x::AbstractString)` -> `first(parse_identifier(x))` (the previous method allowed, but did not return, parent identifiers - and was buggy to boot, allowing invalid parent identifiers)
- general removal of the `schema_` prefix from `SchemaVersion` accessor methods
- `schema_qualified_string` -> `identifier`
- `validate(table, schema)` -> `validate(Tables.schema(table), schema)`

## Where to go from here

Note this PR bumps the version to v0.5, not v1.0, because there are still some obvious breaking changes to be made later:

- to the table-level API (#31, #34)
- to errors that are currently documented as thrown for specific cases (#44)
- get rid of DataFrames dependency and move generic methods elsewhere (e.g. `gather`)
    
My plan is to merge this PR without implementing automated deprecation paths. I think the breakage in this release (esp. w.r.t. `Legolas.Row` removal) is sufficiently nontrivial that any user who'd like to upgrade will probably need to actually manually upgrade, rather than fully rely on automated deprecation pathways. If it's worthwhile, we can add in depwarns for more trivial cases (e.g. simple method renames) in follow-up patch releases, but I'm reluctant to attempt more complex deprecation pathways that would have to handle a lot of edge cases in order to be correct (and thus have a higher risk of being subtly buggy).

Note that Beaconeers aren't expected to upgrade to v0.5 internally without dedicated support from the relevant team members (if you have a question about this, feel free to ping me in Beacon's internal Slack). So feel free to upgrade "early" if you want to for your use case, but keep in mind that it's not a requirement.

### CODE REVIEW

**Comment by hannahilea** [comment]
> More comprehensive introspection/validation utilities [...]

<3

**Comment by jrevels** [comment]
I'm currently working on a branch to upgrade Onda to the intended Legolas v0.5 and running into an area of Legolas' design that is really tempting to try to improve before merging this.

I don't have time to write up a full explanation of the problem space now (will do so later) but suffice to say I'm tempted to reintroduce a struct-generation mechanism that is kind of like `Legolas.Row` but that doesn't allow "extension w/o declaration", i.e. the full field structure is statically known and forcibly encoded in the type definition. either that, or some schema-specialized "table builder" pattern

still unclear on how/whether the field types should be allowed to be parameterized, or whether or not this notion is actually 1:1 w/ schema declaration

will think on it + follow up with a better breakdown of the motivations later

**Comment by jrevels** [comment]
> will think on it + follow up with a better breakdown of the problem space / motivations later

Alrighty then! I jotted down a bunch of thoughts in an attempt to break down the problem/motivations as promised. It's a bit too much content for a single comment so I've dumped the write up in a gist here: https://gist.github.com/jrevels/fdfe939109bee23566d425440b7c759e

I'm going to forge ahead and refocus this PR towards implementing the proposal put forward in that gist. Will flip this into draft mode in the meantime. My hope is to get everything updated and mergeable by EOW.

**Comment by jrevels** [comment]
> My hope is to get everything updated and mergeable by EOW.

Took a little bit longer than expected, but I've now merged https://github.com/beacon-biosignals/Legolas.jl/pull/55 into this PR, which implements the [aforementioned proposal](https://gist.github.com/jrevels/fdfe939109bee23566d425440b7c759e). 

I'm going to try to get CI green, then I'll flip this PR back out of draft mode. My hope is to get this merged/tagged on Thursday, modulo any major concerns brought up in review.

File: src/schemas.jl [hunk 1]
```diff
@@ -0,0 +1,662 @@
...
+
+A type representing a particular version of Legolas schema. The relevant `name` (a `Symbol`)
+and `version` (an `Integer`) are surfaced as type parameters, allowing them to be utilized for
+dispatch.
+
+For more details and examples, please see `Legolas.jl/examples/tour.jl` and the
+"Schema-Related Concepts/Conventions" section of the Legolas.jl documentation.
+
+The constructor `SchemaVersion{name,version}()` will throw an `ArgumentError` if `version` is
+not non-negative.
```
Comment (ericphanson) [inline src/schemas.jl:31]:
```suggestion
negative.
```

File: src/schemas.jl [hunk 2]
```diff
@@ -0,0 +1,662 @@
...
+
+See also: [`Legolas.@schema`](@ref)
+"""
+struct SchemaVersion{n,v}
+    function SchemaVersion{n,v}() where {n,v}
+        v isa Integer && v >= 0 || throw(ArgumentError("`version` in `SchemaVersion{_,version}` must be a non-negative integer, received: $v"))
```
Comment (ericphanson) [inline src/schemas.jl:37]: maybe this is helpful? In case there's a non-`Integer` that prints like an integer:
```suggestion
        v isa Integer && v >= 0 || throw(ArgumentError("`version` in `SchemaVersion{_,version}` must be a non-negative integer, received: $v (of type $(typeof(v))"))
```

File: src/schemas.jl [hunk 3]
```diff
@@ -0,0 +1,662 @@
...
+function Base.showerror(io::IO, e::UnknownSchemaVersionError)
+    print(io, """
+              UnknownSchemaVersionError: encountered unknown Legolas schema version:
+
+                name=\"$(name(e.schema_version))\"
+                version=$(version(e.schema_version))
+
+              This generally indicates that this schema has not been declared (i.e.
+              the corresponding `@schema` and/or `@version` statements have not been
+              executed) in the current Julia session.
```
Comment (ericphanson) [inline src/schemas.jl:112]: if we had an `schema_owner` metadata convention (a la https://github.com/beacon-biosignals/Legolas.jl/issues/46), we could print any such metadata here, to point to what package is missing (e.g.)

File: src/schemas.jl [hunk 4]
```diff
@@ -0,0 +1,662 @@
...
+
+See also: [`Legolas.find_violation`](@ref), [`Legolas.complies_with`](@ref)
+"""
+function validate(ts::Tables.Schema, sv::SchemaVersion)
+    result = find_violation(ts, sv)
+    isnothing(result) && return nothing
+    field, violation = result
+    ismissing(violation) && throw(ArgumentError("could not find expected field `$field` in $ts"))
+    expected = getfield(required_fields(sv), field)
+    throw(ArgumentError("field `$field` has unexpected type; expected <:$expected, found $violation"))
```
Comment (ericphanson) [inline src/schemas.jl:254]: would be nice if we could print all violations if there's more than 1

**Review by ericphanson — COMMENTED** [review]
I think this is great! I think this is a better system than the already-quite-useful `@row` one. I do have fears about a long tail of upgrading old tables, and more specific fears about arrow serialization still not being tested thoroughly (e.g. w/ multiple type parameters, w/ union type constraints, etc), and some more specific comments.


### Dev ergonomicity 1: Revise

Redefining a schema was actually super nice when developing. It made Legolas Row's like Revisable-structs, where you could do
```
@row("x@1", x::Int)
```
and then later realize you actually need a `y` field too, so you just put in the REPL
```
@row("x@1", x::Int, y::Float64)
```
and everything "just works" (since the type is the same and all the info is in the method's which get invalidated).

I get that we don't want to be redefining schemas like this in package code, and don't want to be overwriting existing definitions etc, but... if there's some way to preserve the development ergonomicity, it would be super nice. However I'm not sure how that could actually be implemented now that we are really defining individual structs...

### Dev ergonomicity 2: what was created by this macro

I think one issue with `@version` and `@schema` is that it's not really visible what happened when you run one of those, like what did it generate. `@row` was similar but it returned something, so you had some feedback that OK now I have this object and I can use it. But these return `nothing`, so you don't get a nudge as to what was created and what you can use. (Obvs it's in the docs/tour, but still it's kind of nice to have that automatic feedback).

In particular, the connection between `@version("example.foo@1", ...)` and `FooV1` being created feels too loose (in terms of code clarity, not actual semantics).

One option is `@version` could return say `FooV1`, the type. But I worry this would encourage folks to do `const MyFoo = @version(...)` like we do with `@row`, and that would be kinda an anti-pattern here (since probably it's better to use the FooV1 name rather than introduce another one, unless there is a good reason to).

Another semi weird option would be to try to introspect if the macro is running in the REPL (vs a package/precompiling/etc) and if so, print something like ("FooV1 type has been created"). I think that could go along way to making the package more approachable, even if it's a bit odd.

Or we could somehow try to get `FooV1` into the written code itself, like writing the macro as `@version(FooV1, ...)` and using `@schema` to reverse the connection, meaning that `@version(FooV1, ...)` errors unless an `@schema` has been declared so that FooV1 can be associated to the `example.foo@1` schema. In other words, you wouldn't have any more freedom in choosing how to spell the type (which is good), but you would have to write the type name there, so then you know it exists and when it shows up later in the code you know where it came from.

So that would look like:
```julia
@schema "example.foo" Foo

@version FooV1 begin
    a::Real
    b::String
end
```

**Comment by jrevels** [comment]
> Or we could somehow try to get FooV1 into the written code itself, like writing the macro as `@version(FooV1, ...)` and using `@schema` to reverse the connection, meaning that `@version(FooV1, ...)` errors unless an `@schema` has been declared so that FooV1 can be associated to the example.foo@1 schema. In other words, you wouldn't have any more freedom in choosing how to spell the type (which is good), but you would have to write the type name there, so then you know it exists and when it shows up later in the code you know where it came from.
>
> So that would look like:

Yeah, this does look prettier and clearer from a certain angle

The cons AFAICT: 

- it kind of emphasizes the record type over the schema version type
- it obfuscates the schema version identifier itself a bit
- it would require me to reimplement some stuff 😅 

i think the huge pro, though, is the greppability improvement - even if you don't know what's going on with `FooV1` at a callsite, you can find the `FooV1` definition via text search. i think this is actually pretty important, and i think tips me in favor of this change

i think as part of this change we should also maybe just change `FooSchemaV1` (which reads to me as "foo schema version 1") to `FooV1SchemaVersion` (which reads to me as "the SchemaVersion for FooV1"). Keeping the `V1` next to the `Foo` i think preserves some of the greppability we're getting from . `FooV1Schema` is shorter, I don't like that it kinda conflates "schema" and "schema version", which this PR works pretty hard to cleanly delineate

let the repaint of the bikeshed begin 😁

**Comment by jrevels** [comment]
> I get that we don't want to be redefining schemas like this in package code, and don't want to be overwriting existing definitions etc, but... if there's some way to preserve the development ergonomicity, it would be super nice. However I'm not sure how that could actually be implemented now that we are really defining individual structs...

Yeah, makes sense. I'm not sure how to cleanly support this, but would be down if somebody wants to make a suggestion on how to do so. 

Also worth noting that AFAICT it likely wouldn't be a breaking change if we figured out how to solve this in a subsequent release

File: src/tables.jl [hunk 5]
```diff
@@ -240,3 +97,144 @@ such access costs upfront before repeatedly accessing the table.
...
 materialize(table) = map(collect, Tables.columntable(table))
+
+#####
+##### read/write Arrow content to generic path types
+#####
+# It would be better if Arrow.jl supported a generic API for nonstandard path-like types so that
+# we can avoid potential intermediate copies here, but its documentation is explicit that it only
+# supports `Union{IO,String}`.
+#
+# TODO: upstream improvements to Arrow.jl to obviate these?
```
[the remark this answers is not in the document]
Comment (jrevels) [reply src/tables.jl:108]: Oh, dope. yeah this comment/code exists on `main` as-is, github's diff is just overzealous. we should tackle this TODO in a subsequent release IMO

File: src/schemas.jl [hunk 6]
```diff
@@ -0,0 +1,662 @@
...
+function _validate_wrt_parent(child_fields::NamedTuple, parent::SchemaVersion)
+    declared(parent) || throw(SchemaVersionDeclarationError("parent schema version cannot be used before it has been declared: $parent"))
+    _has_valid_child_field_types(child_fields, required_fields(parent)) || throw(SchemaVersionDeclarationError("declared field types violate parent's field types"))
+    return nothing
+end
+
+function _check_for_expected_field(schema::Tables.Schema, name::Symbol, ::Type{T}) where {T}
+    i = findfirst(==(name), schema.names)
+    if isnothing(i)
+        Missing <: T || return missing
```
[the remark this answers is not in the document]
Comment (jrevels) [reply src/schemas.jl:368]: yup, the one caller of this function passes in `accepted_field_type(sv, S)` as the `T` argument

File: src/schemas.jl (hunk 1 above)
Comment (jrevels) [reply src/schemas.jl:31]:
```suggestion
not not non-non-negative.
```

?

😁

File: examples/tour.jl [hunk 7]
```diff
@@ -3,192 +3,392 @@
...
+
+# Eliding the required field `c`, which is assigned `missing` in the output:
+foo = FooV1(; a=1.0, b="hi", d=[1, 2, 3])
+@test isequal(NamedTuple(foo), (a=1.0, b="hi", c=missing, d=[1, 2, 3]))
+
+# Providing the non-compliantly-typed field `d::Int`, inducing a `MethodError`:
+@test_throws MethodError FooV1(; a=1.0, b="hi", d=2)
+
+# Implicitly providing the non-compliantly-typed field `d::Missing`, inducing a `MethodError`:
+@test_throws MethodError FooV1(; a=1.0, b="hi")
```
[the remark this answers is not in the document]
Comment (jrevels) [reply examples/tour.jl:121]: yeah, this is basically https://github.com/beacon-biosignals/Legolas.jl/issues/44

I'll give it a small ponder to see if this can be fixed easily in this iteration of the PR. if not, though, i think i'll leave tackling that issue to a subsequent release

File: examples/tour.jl [hunk 8]
```diff
@@ -3,192 +3,392 @@
...
+#####
+##### `Tables.Schema` Compliance/Validation
+#####
+
+# We can use `complies_with`, `validate`, and `find_violation` to check whether a given
+# `Tables.Schema` (ref https://tables.juliadata.org/stable/#Tables.Schema) complies with
+# `example.foo@1`.
+
+# For example, all of the following `Tables.Schema`s comply with `example.foo@1`:
+for s in [Tables.Schema((:a, :b, :c, :d), (Real, String, Any, AbstractVector)), # All required fields must be present...
```
[the remark this answers is not in the document]
Comment (jrevels) [reply examples/tour.jl:50]: agreed!

[the remark this answers is not in the document]
Comment (jrevels) [reply examples/tour.jl:50]: xref https://github.com/JuliaData/Tables.jl/issues/314

File: src/schemas.jl (hunk 4 above)
Comment (jrevels) [reply src/schemas.jl:254]: Yeah. I'm fine leaving that to a subsequent release though. feel free to file an issue

Comment (ericphanson) [reply src/schemas.jl:254]: https://github.com/beacon-biosignals/Legolas.jl/issues/56

File: test/runtests.jl [hunk 9]
```diff
@@ -109,93 +108,259 @@ end
...
+        meta
+    end
+    Legolas.Arrow.getmetadata(moo::Moo) = moo.meta
+    moo = Moo(Dict("a" => "b", "b" => "b", Legolas.LEGOLAS_SCHEMA_QUALIFIED_METADATA_KEY => "test.parent@1"))
+    @test ParentSchemaV1() == Legolas.extract_schema_version(moo)
 
-@testset "Nested arrow serialization" begin
-    table = [MyOuterRow(; a="outer_a", x = MyInnerRow())]
-    roundtripped_table = Legolas.read(Legolas.tobuffer(table, Legolas.Schema("my-outer-schema@1")))
-    @test table == MyOuterRow.(Tables.rows(roundtripped_table))
```
[the remark this answers is not in the document]
Comment (jrevels) [reply test/runtests.jl:200]: yes! good catch, this omission was accidental

**Comment by hannahilea** [comment]
> if we figured out how to solve this

...could always poke the relevant upstream issue (Revise+struct) folks for this one? https://github.com/JuliaLang/julia/issues/40399 via https://github.com/timholy/Revise.jl/issues/18#issuecomment-816589248

File: examples/tour.jl [hunk 10]
```diff
@@ -3,192 +3,392 @@
...
+#       y::String = string(y, '_', x)
+#       return new(x, y)
+#   end
 #
-# Here we demonstrate the first property - required fields have a `missing` value by default:
-@test isequal(MyRow(a=1.5, b="hello", c="goodbye", d=2), MyRow((; a=1.5, b="hello", c="goodbye", d=2, e=missing)))
-@test_throws MethodError MyRow(a=1.5, b="hello", c="goodbye") # correctly throws a `MethodError` when evaluating `d::Int = missing`
+# ...so that invocations `BarV1(; ...)` have the following behavior:
+@test NamedTuple(BarV1(; x=200, y=:hi)) == (x=127, y="hi", z="hi_127")
+@test isequal(NamedTuple(BarV1(; y=:hi)), (x=missing, y="hi", z="hi_missing"))
```
[the remark this answers is not in the document]
Comment (jrevels) [reply examples/tour.jl:149]: (this might align exactly with what you meant, but just for clarity) 

> which means that you can't add a new non->:Missing column

the table itself can have a non-`>:Missing` non-required column and still be compliant, but adding a new `non->:Missing` required field to the schema would be a breaking change that would entail a new schema version 

(which is no different than the current Legolas release)

File: examples/tour.jl [hunk 11]
```diff
@@ -3,192 +3,392 @@
...
+#       z::String = z
+#       k::Int = length(z)
+#       return new(x, y, z, k)
+#   end
+
+# One last note on syntax: You might ask "Why use `>` as the inheritance operator instead of `<:`?" There are two reasons.
+# The primary reason is purely historical: earlier versions of Legolas did not as rigorously demand/enforce subtyping
+# relationships between parent and child schemas' required fields, and so the `<:` operator was considered to be a bit
+# too misleading. A secondary reason in favor of `>` was that it implied the actual order of application of field
+# constraints/transformations (i.e. the parent's are applied before the child's).
```
[the remark this answers is not in the document]
Comment (jrevels) [reply examples/tour.jl:244]: yup, this is how i think of it too, the child has more constraints than the parent

[the remark this answers is not in the document]
Comment (jrevels) [reply examples/tour.jl]: https://github.com/beacon-biosignals/Legolas.jl/pull/54/commits/5f8ec57574cbe8d8372dc165705e16c6390077b2

File: test/runtests.jl (hunk 9 above)
[the remark this answers is not in the document]
Comment (jrevels) [reply test/runtests.jl:200]: https://github.com/beacon-biosignals/Legolas.jl/pull/54/commits/2fe38fc82ef3f2a9fc92622cd86cf811c127369c

File: examples/tour.jl (hunk 7 above)
[the remark this answers is not in the document]
Comment (jrevels) [reply examples/tour.jl:121]: Okay, ponder has been ponder'd. 

The path to implementing this is a bit intertwined with [this line of thinking](https://github.com/beacon-biosignals/Legolas.jl/pull/54#discussion_r1005849968), funnily enough. Basically, we'd only want to perform the kind of validation that gives us the nice error message *after* construction itself has been attempted, since the inputs are not necessarily expected to de facto be valid outputs. But it's the construction bit that throws the poor error in the first place.

this implies the route to surfacing a better error message is to generate a `try/catch` like this:

```jl
# inner constructor
function FooV1(; x=missing, y=missing)
    try
       x::X = ...
       y::Y = ...
    catch err
        @warn "<better descriptive message, possibly using find_violation>"
        rethrow(err)
    end
    return new(x, y)
end
```

not too bad, but this is just enough extra work that i'd rather tackle #44 in a future PR. will xref this from #44

**Comment by jrevels** [comment]
> let the repaint of the bikeshed begin

Okay, bikeshed repainted. Retagging @ericphanson for one last review pass 🙏 

If there are no major concerns my plan is to merge by EOD tomorrow and tag the v0.5 release as stated in the OP.

File: docs/src/schema-concepts.md [hunk 12]
```diff
@@ -13,13 +13,13 @@ Legolas defines "schema version identifiers" as strings of the form:
...
 
 Schema authors should follow the below conventions when choosing the name of a new schema:
 
 1. Include a namespace. For example, assuming the schema is defined in a package Foo.jl, `foo.automobile` is good, `automobile` is bad.
 2. Prefer singular over plural. For example, `foo.automobile` is good, `foo.automobiles` is bad.
-3. Don't "overqualify" the schema name with ancestor-derived information. For example, `bar.automobile@1>foo.automobile@1` is good, `baz.supercar@1>bar.automobile@1` is good, `bar.foo.automobile@1>foo.automobile@1` is bad, `baz.automobile.supercar@1>bar.automobile@1` is bad.
+3. Don't "overqualify" a schema name with ancestor-derived information that is better captured by the fully qualified identifier of a specific schema version. For example, `bar.automobile` should be preferred over `bar.foo.automobile`, since `bar.automobile@1>foo.automobile@1` is preferrable to `bar.foo.automobile@1>foo.automobile@1`. Similarly, `baz.supercar` should be preferred over `baz.automobile.supercar`, since `baz.supercar@1>bar.automobile@1` is preferrable to `baz.automobile.supercar@1>bar.automobile@1`.
```
Comment (ericphanson) [inline docs/src/schema-concepts.md:22]:
```suggestion
3. Don't "overqualify" a schema name with ancestor-derived information that is better captured by the fully qualified identifier of a specific schema version. For example, `bar.automobile` should be preferred over `bar.foo.automobile`, since `bar.automobile@1>foo.automobile@1` is preferable to `bar.foo.automobile@1>foo.automobile@1`. Similarly, `baz.supercar` should be preferred over `baz.automobile.supercar`, since `baz.supercar@1>bar.automobile@1` is preferrable to `baz.automobile.supercar@1>bar.automobile@1`.
```

**Review by ericphanson — APPROVED** [review]
Looks good to me. I would suggest we also document some update guidance & best practices, maybe in other PRs (I know you said everyone will get help upgrading but I think it still helps to have some written guidance). E.g., questions I have would be:

* when should we be making a schema extension vs using custom columns (given constructors drop custom cols)?
    * Should custom cols only be used when it's entirely dynamic what the column is, so it's not really possible to write a schema unless we generate it on-the-fly (which sounds like a very bad idea), or are custom columns more broadly applicable?
* since we can only check type-constraints (not value constraints) without the constructors, whose responsibility is it to do that construction? the writer or the reader or both?
    * I am guessing the reader, or maybe more specifically, the caller of a function requiring that vetting, since you shouldn't really trust the writer did it, since who knows what they did
    * if it is the reader/caller, then this also means it's possible to serialize out custom columns, otherwise they would be dropped by constructing-before-writing, and you can read them in as long as you don't do that constructing
* is it more OK to constrain function signatures by type now than before? E.g. if you want to express that you need a constructed-therefore-validated type as input
* is it more OK to dispatch on record types now than before? Since your package "owns" the type rather than Legolas now, and they seem to have greater semantic meaning now
    * this is kinda the same as the previous question, but imo relying on dispatch for functionality is different semantically than e.g. having a function w/ a single method w/ a type constraint

also, unrelatedly, if we want table-level validation, and we are in the validate-by-construction framework, does that mean we need tabular types, not just row types?

**Comment by jrevels** [comment]
> when should we be making a schema extension vs using custom columns (given constructors drop custom cols)?

i'm not sure there's a hard rule, so i'd just say "when it's useful to do so" (which i know is annoying 😁 but maybe we can figure out a more thoughtful rule at some point). one's use of schema-aware tools/systems may incentivize this in different directions

> since we can only check type-constraints (not value constraints) without the constructors, whose responsibility is it to do that construction? the writer or the reader or both?

For the reader, it's all about trust; for the writer, it's all about goals. A big part of this is whether you're reading from / writing to a schema-on-read system or a (trusted) schema-on-write system. For example, if I'm a reader that

i have some design notes written down on this topic that i'd like to put into a design docs section of the docs

> is it more OK to constrain function signatures by type now than before? E.g. if you want to express that you need a constructed-therefore-validated type as input

Yup, it is; but with the caveat that doing so also means biting the bullet that your code is now non-generic, and thus could "poison" otherwise generic code that callers might want to compose it with. which, in the worst case, could result in costly/burdensome conversion gymnastics (but in some cases, that might be okay)

> is it more OK to dispatch on record types now than before? Since your package "owns" the type rather than Legolas now, and they seem to have greater semantic meaning now

Yeah, but maybe less so than the former one lol. 

Basically it's okay to do these things if you understand/accept the consequences. i personally think a more duck-typed style is more useful/less frustrating in a lot of scenarios, given the plethora of types that a valid table/row value can take on. IMO the holy grail would be true structural typing support, but that might be a bit of a lofty non-goal in julia land

> also, unrelatedly, if we want table-level validation, and we are in the validate-by-construction framework, does that mean we need tabular types, not just row types?

either that, or have tabular validation be treated separately. i actually explored introducing a table type in this PR a little bit before going with the record type approach

> update guidance & best practices

agreed, i'll file an issue for this
